pkg/nmi/server/server.go

```diff
@@ -380,3 +384,57 @@ func handleTermination() {
 	log.Infof("Exiting with %v", exitCode)
 	os.Exit(exitCode)
 }
+
+func listPodIDsWithRetry(ctx context.Context, kubeClient k8s.Client, logger *log.Entry, podns, podname, rqClientID string, maxAttemptsPerCheck int) ([]aadpodid.AzureIdentity, error) {
+	attempt := 0
+	var err error
+	var idStateMap map[string][]aadpodid.AzureIdentity
+
+	// this loop will run to ensure we have assigned identities before we return. If there are no assigned identities in created state within 30s, then we return an error.
+	// If we get an assigned identity in created state within 30s, then loop will continue for another 30s to find assigned identity in assigned state.
+	for attempt < 2*maxAttemptsPerCheck {
+		idStateMap, err = kubeClient.ListPodIds(podns, podname)
+		if err == nil {
+			if len(rqClientID) == 0 {
+				// check to ensure backward compatability with assignedIDs that have no state
+				if len(idStateMap[""]) != 0 {
+					logger.Warningf("found assignedIDs with no state for pod:%s/%s. AssignedIDs created with old version of mic.", podns, podname)
+					return idStateMap[""], nil
+				}
+				if len(idStateMap[aadpodid.AssignedIDAssigned]) != 0 {
+					return idStateMap[aadpodid.AssignedIDAssigned], nil
+				}
+			}
+
+			// if client id exists in request, we need to ensure the identity with this client
+			// exists and is in Assigned state
+			if len(rqClientID) != 0 {
+				// check to ensure backward compatability with assignedIDs that have no state
+				for _, podID := range idStateMap[""] {
+					if strings.EqualFold(rqClientID, podID.Spec.ClientID) {
+						logger.Warningf("found assignedIDs with no state for pod:%s/%s. AssignedIDs created with old version of mic.", podns, podname)
+						return idStateMap[""], nil
+					}
+				}
+				for _, podID := range idStateMap[aadpodid.AssignedIDAssigned] {
+					if strings.EqualFold(rqClientID, podID.Spec.ClientID) {
+						return idStateMap[aadpodid.AssignedIDAssigned], nil
+					}
+				}
+			}
+			if len(idStateMap[aadpodid.AssignedIDCreated]) == 0 && attempt >= maxAttemptsPerCheck {
+				return nil, fmt.Errorf("getting assigned identities for pod %s/%s in CREATED state failed after %d attempts. Error: %v", podns, podname, maxAttemptsPerCheck, err)
+			}
+		}
+		attempt++
+
+		select {
+		case <-time.After(listPodIDsRetryIntervalInSeconds * time.Second):
+		case <-ctx.Done():
+			err = ctx.Err()
+			return nil, err
+		}
+		logger.Warningf("failed to get assigned ids for pod:%s/%s in ASSIGNED state, retrying attempt: %d", podns, podname, attempt)
+	}
+	return nil, fmt.Errorf("getting assigned identities for pod %s/%s in ASSIGNED state failed after %d attempts. Error: %v", podns, podname, 2*maxAttemptsPerCheck, err)
```

Comment:
It would be good to have the duration in the error logs on how much it tried to do retry.